### DISCUSSION TITLE
Expressions returning only `Just _ : Maybe a` -> convert to return `a` instead

### DISCUSSION THREAD

**Discussion by Janiczek** [jfmengels/elm-review#89] [Rule ideas]
**What the rule should do:**
Check all returned expressions inside `if...else` and `case...of` expressions of type `Maybe _` for whether they're all `Just`s.
In that case it should suggest that you change it to instead return the values inside the `Just`s and (perhaps?) call the `Just` constructor outside (or refactor the usages of this expression/function to something nicer / more idiomatic).

**Example of things the rule would report:**
```elm
case stuff of
  X -> Just "x"
  Y -> Just "y"
  Z -> 
    if foo then
      Just <| blabla "zzz" 999
    else
      Just <| somethingMoreComplicated "z" 42
```
:arrow_down:
```elm
case stuff of
  X -> "x"
  Y -> "y"
  Z -> 
    if foo then
      blabla "zzz" 999
    else
      somethingMoreComplicated "z" 42
```

or

```elm
if foo then
  Just "x"
else
  Just "y"
```
:arrow_down:
```elm
if foo then
  "x"
else
  "y"
```

**Example of things the rule would not report:**
```elm
if foo then
  Just "x"
else
  Nothing
```

**When (not) to enable this rule:**

Libraries might be bound by their existing API design and so it might not be desirable to refactor such functions to stop exposing `Maybe`.

**I am looking for:**
- Feedback
- Someone to implement it with/for me

**Comment by MartinSStewart** [comment]
If this rule is just checking that all the branches in an expression end with the same custom type variant, maybe it could be generalize to handle other Ok, Err, and user defined custom type variants as well?
